crypto/x509/x509_vfy.c

```diff
@@ -494,23 +494,27 @@ static int check_chain(X509_STORE_CTX *ctx)
                          && ret != 1), ctx, x, i, X509_V_ERR_INVALID_CA);
             break;
         }
-        if (num > 1) {
-            /* Check for presence of explicit elliptic curve parameters */
-            ret = check_curve(x);
-            CHECK_CB(ret < 0, ctx, x, i, X509_V_ERR_UNSPECIFIED);
-            CHECK_CB(ret == 0, ctx, x, i, X509_V_ERR_EC_KEY_EXPLICIT_PARAMS);
-        }
+
         /*
-         * Do the following set of checks only if strict checking is requested
-         * and not for self-issued (including self-signed) EE (non-CA) certs
+         * Skip further checks for self-issued (including self-signed) EE certs
          * because RFC 5280 does not apply to them according RFC 6818 section 2.
          */
-        if ((ctx->param->flags & X509_V_FLAG_X509_STRICT) != 0
-            && num > 1) { /*
-                           * this should imply
-                           * !(i == 0 && (x->ex_flags & EXFLAG_CA) == 0
-                           *          && (x->ex_flags & EXFLAG_SI) != 0)
-                           */
+        if (i == 0 && (x->ex_flags & EXFLAG_CA) == 0
+                && (x->ex_flags & EXFLAG_SI) != 0)
+            continue;
+
+        /* Reject unknown critical extensions due to RFC 5280 section 4.2 */
```

Comment:
Why are you moving this check down here? Ignoring unknown critical extensions even on root certs does not seem to me to be right. Although this i relaxing what we have now and so it should not break existing use-cases I do not see the need for the relaxing.

Reply:
IMO doing such checks on certs that are explicitly exempted(!) from RFC requirements should not be done by default, but (if at all) only when being strict.
In particular in this case where it's about unknown extensions. 
When such extensions are placed in certs used as trust anchors, where most of their contents are clearly irrelevant, it could be even harmful to forcefully reject any extensions unknown to OpenSSL.

Reply:
OK

Reply:
I would like to hear @vdukhovni 's thoughts on this one.
It seems to touch on the question of whether a certificate is just a convenience encoding for a trust anchor key, or one expects to also apply semantic value from the other parts of the certificate encoding; to my knowledge, this question does not have a single authoritative answer.  We may be in the regime where people have come to rely on whatever undocumented behavior openssl has historically implemented.  So, I personally am wary of much change in this area (but I am not putting a -1 on the PR).
(I am cognizant of Viktor's previous remarks about X509_AUX, etc. and that we may even want to discourage the use of the certificate fields themselves as applying metadata to the trust anchor key.  But I don't know either way.)